### ISSUE TITLE
Formations - Cannot save roster

### ISSUE THREAD

**Issue by RBarrDev13** [BSData/wh40k-7th-edition#131]
If you use any of the Ork Formations, such as Da Vulcha Skwad or the Ork Warband, and attempt to save the roster file, you get the following error for each unit in the formation:

"Selection must have a category that exists"

Steps to replicate:

1) Create a new roster that uses a Formation Detachment from Codex: Orks (2014)
2) Add a Formation.
3) Save the file.

Expected behavior: file saves normally.
Actual behavior: save fails, "Selection must have a category that exists" error line appears for each unit in the formation.

**Comment by Kangodo** [comment]
#132 has the same issue with Tyranids. Probably related?

**Comment by HeWhoWatches** [comment]
TLDR: HQ, Elites, etc. categories don't exist under Formation Detachment in the Game System? 
Think this is because the categories assigned (e.g. "HQ" for Zagstruk) do not exist under the Formation Detachment when you load the 40k Game System. If you simplify the entry in the Ork catalogue by removing everything but Zagstruk you can still replicate the problem. Then if you load the CAT file in a text editor, you can see more easily which category ID he is associated with (848a6ff2-0def-4c72-8433-ff7da70e6bc7 = HQ). If you manually replace the UID with the only available option under the Formation Detachment in the 40k Game System (28b94f51-e66b-4096-aa59-0c9df620a77d = Formation), then try to create a list with the formation and save it, it no longer errors.

**Comment by amis92** [comment]
Oh yes, long ago I had the same problem with Tau catalogue.

The problem is actually in BattleScribe (@Jonskichov), because we cannot help it on the Catalogue Editor level.

The problem is as follows:
- When choosing Force Type `Formation Detachment`, BScribe reads only category `Formation`.
- Then we choose an entry of Formation we want. It's all right because BScribe probably reads whole gst file anyway (and so knows all categories).
- When we save - BAM - we can't because units _inside_ the formation have actually been assigned a category (which isn't good in the first place, but let me go) like HQ or Troops, which _isn't_ to be included/mentioned in XML file BScribe wants to generate, and so _theoretically_ would cause thhis file to be unreadable by BScribe when loaded.

The problem arises when we copy/paste units/entries from root level into shared level (or whatever). Because all root level entries have to have a category, the entry we paste _has_ category, which isn't needed except on root level.

TLDR/Summary: you need to edit XML file, and replace the categoryId of entries which are included in formation with "(No Category)". Actually, it would be best, if you replaced _all_ categoryIds _except_ of those on root level, where they are actually used.

In few hours I'll be able to write up an XSLT to do exactly that automatically. Using my BSReshaper, of course ;)

So, patience please :D

**Comment by ghost** [comment]
Edited the Orks file in xml, but I'd still appreciate that xslt @amis92, because the others will need doing.

**Comment by amis92** [comment]
Little update @khambatta I hadn't had enough time before my departure to a place disconnected from the Internet :)

Now I can write it. Will do it in next few days, most probably tomorrow.

**Comment by amis92** [comment]
Ok now I see BS will disallow it, so I'll wait if it doesn't fix the problem from get-go. :)
(link: https://plus.google.com/112844273639962659291/posts/NT4nZt77PWp).

**Comment by ghost** [comment]
Once I figured out what I was doing, I edited the xml in the files I've added formations to, so they're sorted. It was just a case of replacing the relevant units' category ID with the "[No Category]" tag - a lot simpler than I had feared.
I don't know anything about the 'Nids, or other people's work, though.

**Comment by Randobar** [comment]
Did the "[No Category]" fix for the Tyranids.

**Comment by Kangodo** [comment]
Could I ask how you solved this issue? I don't exactly understand the solution.
I found out that the Necron-codex has the same issue while I was preparing for an upcoming Apocalypse.

**Comment by Randobar** [comment]
like khambatta said. I extracted the .cat from the catz, opened it with a texteditor and changed the category of all units inside formations to "[No Category]".

I suggest an editor that i capable to unterstand XML and therefore enables you to keep track of the entrys and brackets.

**Comment by amis92** [comment]
@Randobar - Ok, so to solve the issue, follow these steps:
1. Download BSReshaper.exe from [here](https://github.com/amis92/bs-xml-reshaper/releases/latest)
2. Copy the catalogue into some new empty folder (in case something goes wrong, and because BSReshaper transforms all .cat and .gst files in chosen folder).
3. Save somewhere this as something.xsl

``` xsl
<?xml version="1.0" encoding="UTF-8"?>
<xsl:stylesheet version="1.0" xmlns:xsl="http://www.w3.org/1999/XSL/Transform"
xmlns:bs-cat="http://www.battlescribe.net/schema/catalogueSchema">

  <xsl:output omit-xml-declaration="no" encoding="UTF-8" standalone="yes" version="1.0" indent="yes" method="xml"/>

  <!-- The Identity Transformation -->
  <!-- Whenever you match any node or any attribute -->
  <xsl:template match="node()|@*">
    <!-- Copy the current node -->
    <xsl:copy>
      <!-- Including any attributes it has and any child nodes -->
      <xsl:apply-templates select="@*|node()"/>
    </xsl:copy>
  </xsl:template>

  <!-- Removing categoryId from entry descendants of any level of sharedEntries node-->
  <xsl:template match="bs-cat:sharedEntries//bs-cat:entry/@categoryId">
    <xsl:attribute name="{name()}" namespace="{namespace-uri()}">(No Category)</xsl:attribute>
  </xsl:template>

  <!-- Removing categoryId from entry descendants of any level of sharedEntryGroups node-->
  <xsl:template match="bs-cat:sharedEntryGroups//bs-cat:entry/@categoryId">
    <xsl:attribute name="{name()}" namespace="{namespace-uri()}">(No Category)</xsl:attribute>
  </xsl:template>

  <!-- Removing categoryId from entry descendants of at least second level in entries node-->
  <xsl:template match="//bs-cat:entries/bs-cat:entry//bs-cat:entry/@categoryId">
    <xsl:attribute name="{name()}" namespace="{namespace-uri()}">(No Category)</xsl:attribute>
  </xsl:template>

</xsl:stylesheet>
```
1. Open BSReshaper.exe, choose the xsl file you've just saved, choose the folder with catalogue you want to repair and click start.
2. After a moment it will be ready (it prints some info).
3. Open the catalogue in BattleScribe cat editor, increment revision number and save.
4. Copy the catalogue back into wh40k folder.
5. Check if you can build and save the roster with formations.
6. Commit changes.
7. Sync them with server.

**Comment by Randobar** [comment]
since this is a general problem with the latest version of BS we should open one issue for each affected cat and keep them seperated, otherwise we wont be able to close this.

**Comment by amis92** [comment]
No reason to blow it up out of proportions. I'll just do it.

**Comment by Kangodo** [comment]
Thank you @amis92 , worked perfectly.

**Comment by amis92** [comment]
Catalogues with _any_ Formations, checked work (can be saved):
- [x] Astra Militarum: Codex
- [x] Chaos Space Marines: Codex
- [x] Chaos Space Marines: Supplement - Crimson Slaughter
- [x] Eldar: Codex
- [x] Militarum Tempestus: Codex
- [x] Necrons: Codex
- [x] Orks: Codex
- [x] SM Supplement - Clan Raukaan
- [x] SM Supplement - Sentinels of Terra
- [x] Space Wolves: Codex
- [x] Tau Empire: Codex
- [x] Tau Empire: Supplement - Farsight Enclaves
- [x] Tyranids: Codex
